### PR TITLE
Prodreq 35124/aws sdk2.0 sqs/initial setup

### PR DESCRIPTION
**Pull request by willianpc**
WIP: Just to have the PR going through the pipeline

### CODE REVIEW

File: packages/collector/package.json
Comment (basti1302) [inline packages/collector/package.json:123]: 🎉 👍 👍 🎉

File: packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js:57]: Can `data` contain something like `data.error`? I think I saw something like that in the instrumentation. We should handle that case here. Maybe just writing the error to the console, so it doesn't go unnoticed.

File: packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js:56]: This then has no catch. It probably should have one.

File: packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js:81]: I guess we could omit this catch I guess and just have the outer then have a catch. Since we return this inner promise, an error in this inner promise should bubble up to the outer promise I think.

File: packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js:90]: similar to the promise case, this await is not wrapped in try catch. We probably should do that. WDYT?

OTOH, an error here would be caught in pollForMessages, that might be good enough

**Review by basti1302 — CHANGES_REQUESTED** [review]
I went through the changes for a detailed review. 

In addition to the individual comments, here are a few things that need to be done before this can be merged:
- [x] more test scenarios for the integration test:
    - [x] tracing disabled
    - [x] fix CI with queue name regex
    - [x] tracing supression
    - [x] with error in sender 
    - [x] with error in receiver
    - [ ] instrument sendMessageBatch()
- [ ] manual test with a Java SQS app and/or Go
- [ ] capture the additional attributes in the SQS span we identified
- [x] check why CI is failing (my guess is because we didn't yet provide AWS credentials)
- [ ] we should also squash all changes into a single commit IMO

File: packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js
Comment (willianpc) [reply packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js:90]: I think that a try/catch wrap is a good idea. We already have a try/catch a few lines below, which doesn't make a lot of sense to have more than one try/catch combo in the same function, IMO.

File: packages/collector/test/tracing/cloud/aws/sqs/sendMessage.js
Comment (willianpc) [reply packages/collector/test/tracing/cloud/aws/sqs/sendMessage.js:75]: In the very beginning, I wasn't sure if there could be any difference between calling `.promise().then` and doing this in two steps by defining a variable with `.promise()` then using it. 
It doesn't make any difference (as expected) and is not needed. I will remove it.

File: packages/collector/test/tracing/cloud/aws/sqs/sqsUtil.js
Comment (willianpc) [reply packages/collector/test/tracing/cloud/aws/sqs/sqsUtil.js:12]: I ran the tests without this option and it worked just fine.
This option was present in [AWS SQS official example code](https://docs.aws.amazon.com/code-samples/latest/catalog/javascript-sqs-sqs_deletequeue.js.html), so I just kept it. They don't mention in the example code if this is mandatory. So I guess we can remove it. 
Any particular reason why you are asking?

Comment (basti1302) [reply packages/collector/test/tracing/cloud/aws/sqs/sqsUtil.js:12]: Uhm, no, actually I was just curious.

I could imagine that the AWS SDK will just use a default apiVersion (maybe the most recent one) if none is provided. It is probably quite irrelevant.

Comment (willianpc) [reply packages/collector/test/tracing/cloud/aws/sqs/sqsUtil.js:12]: Ok. So I removed it and tested against the real SQS and it just works. So I will keep the constructor free of any arguments.

File: packages/core/src/tracing/instrumentation/cloud/aws/sqs.js
Comment (willianpc) [reply packages/core/src/tracing/instrumentation/cloud/aws/sqs.js:136]: Got it. This was copied from GCP, where there is no span.cancel() in the suppression bit. So maybe we should take a look there as well?
https://github.com/instana/nodejs-sensor/blob/main/packages/core/src/tracing/instrumentation/cloud/gcp/pubsub.js#L85

Comment (basti1302) [reply packages/core/src/tracing/instrumentation/cloud/aws/sqs.js:136]: Huh. You are right. Very good catch! 

For pubsub there is already a test that makes sure we do not create a span when suppression is on. And thinking about it, I now see why it works: There is a check for the parent span a few lines later (here in your code as well as in GCP pubsub). And that check will fail when suppression is on because the entry instrumentation (for example httpServer.js) will not have created a span. 

Thinking about it, we should be more explicit about this and not rely on the behaviour of the entry instrumentation. I checked other exit instrumentation (amqp, kafkaJs, httpClient) and we are not entirely consistent but most explicitly stop the processing by doing something equivalent to `return originalSendMessage.apply(ctx, originalArgs);` when suppression is detected. 

In conclusion, we should add the line here and also for pubsub. It won't change the behaviour in practice but make the code easier to understand and more explicit.

File: packages/collector/test/tracing/cloud/aws/sqs/test.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/test.js:39]: I use `.only` a lot locally, but usually try to not commit it. `eslint` will fail due to it. (There is rule that forbids only so it is not accidentally committed)

File: CHANGELOG.md
Comment (basti1302) [inline CHANGELOG.md:5]: Please add full stop at the end. Yes, I am very German and nitpicky, thanks for asking :-)

File: packages/core/src/tracing/instrumentation/protocols/httpClient.js
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/protocols/httpClient.js:72]: Please remove the TODO comment. It has been resolved :-)

File: packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/receiveMessage.js:176]: I think `/* data */` could be removed.

File: packages/collector/test/tracing/cloud/aws/sqs/test.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/test.js:49]: The `only` needs to be removed.

File: packages/core/src/tracing/instrumentation/cloud/aws/sqs.js
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/cloud/aws/sqs.js:229]:
```suggestion
          `Cannot start an AWS SQS entry span when another span is already active. Currently, the following span is active: ${JSON.stringify(
```

File: CHANGELOG.md
Comment (basti1302) [inline CHANGELOG.md:5]:
```suggestion
- Introduction of AWS SQS instrumentation for sending and reading messages.
```

File: packages/collector/test/tracing/cloud/aws/sqs/test.js
Comment (basti1302) [inline packages/collector/test/tracing/cloud/aws/sqs/test.js:49]:
```suggestion
mochaSuiteFn('tracing/cloud/aws/sqs', function() {
```

**Review by basti1302 — APPROVED** [review]
That looks good! 🎉 🎉 

Some very minor nitpicks, see inline comments.

**Pull request merged**
